lib/benchee/statistics/percentile.ex

```diff
@@ -1,50 +1,67 @@
-  5.0
-
-  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, 150)
-  5.0
+  Passing a number for `percentile_rank` calculates a single percentile.
+  Passing a list of numbers calculates multiple percentiles, and returns them
+  as a map like %{90 => 45.6, 99 => 78.9}, where the keys are the percentile
+  numbers, and the values are the percentile values.
 
-  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, 0)
-  1.0
+  ## Examples
 
-  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, -1)
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 12.5)
   1.0
 
   iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 50)
   3.0
 
   iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 75)
   4.75
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 99)
+  5.0
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], [50, 75, 99])
+  %{50 => 3.0, 75 => 4.75, 99 => 5.0}
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 100)
+  ** (ArgumentError) percentile must be between 0 and 100, got: 100
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 0)
+  ** (ArgumentError) percentile must be between 0 and 100, got: 0
   """
-  @spec percentile(list(number()), integer()) :: float()
-  def percentile(samples, percentile_number) do
-    percentile(samples, length(samples), percentile_number)
+  @spec percentile(list(number()), number() | list(number())) ::
+    percentile | percentiles
+  def percentile(samples, percentile_ranks) do
+    number_of_samples = length(samples)
+    sorted = Enum.sort(samples)
+    percentile(sorted, number_of_samples, percentile_ranks)
   end
 
-  @spec percentile(list(number()), integer(), integer()) :: float()
-  def percentile(samples, number_of_samples, percentile_number) when percentile_number > 100 do
-    percentile(samples, number_of_samples, 100)
+  defp percentile(sorted_samples, number_of_samples, percentile_ranks)
+    when is_list(percentile_ranks) do
+    percentile_ranks
+    |> Enum.reduce(%{}, fn percentile_rank, acc ->
+      perc = percentile(sorted_samples, number_of_samples, percentile_rank)
+      Map.put(acc, percentile_rank, perc)
+    end)
   end
 
-  def percentile(samples, number_of_samples, percentile_number) when percentile_number < 0 do
-    percentile(samples, number_of_samples, 0)
+  defp percentile(_, _, percentile_rank) when percentile_rank >= 100 or percentile_rank <= 0 do
+    raise ArgumentError, "percentile must be between 0 and 100, got: #{inspect(percentile_rank)}"
   end
 
-  def percentile(samples, number_of_samples, percentile_number) do
-    sorted = Enum.sort(samples)
-    rank = (percentile_number / 100) * max(0, number_of_samples + 1)
-    percentile_value(sorted, rank)
+  defp percentile(sorted_samples, number_of_samples, percentile_rank) do
```

Comment:
these days we usually group the same function together without a newline inbetween, should find/create a credo rule for that :)

Reply:
@devonestes code formatter now separates these functions, no?

Reply:
what... everything I know is a lie then I believe I started adopting that style after seeing how it's done in elixir core :| I probably should play with the formatter a little...

Reply:
Yep, there's always a space between functions defined with the do-newline-end syntax. If you want no line between functions, you can use `, do:` in the pure AST form.

Reply:
:'(

Reply:
@PragTob yes, some of the code formatter styles are sad, I agree.